### PR TITLE
Custom messages per Build result

### PR DESCRIPTION
**Pull request by KreAch3R**
Following #169 #385 

I re-wrote this patch for the current master source.

While I'm there, sync the notify booleans with the custom message boxes.

Closes https://github.com/jenkinsci/slack-plugin/issues/163
Closes https://github.com/jenkinsci/slack-plugin/issues/360
Closes https://github.com/jenkinsci/slack-plugin/issues/284
Closes https://github.com/jenkinsci/slack-plugin/issues/212

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -229,6 +229,13 @@
                     <compatibleSinceVersion>2.0</compatibleSinceVersion>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <configuration>
+                    <forkCount>0</forkCount>
```
Comment (timja) [inline pom.xml]: what's this needed for?

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java [hunk 2]
```diff
@@ -443,9 +443,21 @@ public MessageBuilder appendFailedTests() {
             return this;
         }
 
-        public MessageBuilder appendCustomMessage() {
+        public MessageBuilder appendCustomMessage(Result buildResult) {
             String customMessage = notifier.getCustomMessage();
-
+            if (buildResult != null) {
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: Thanks @KreAch3R 
This looks great

One problem though imo is that this is a breaking change.

Currently when someone uses the custom message it applies to all statuses

With this change it only applies to the `Notify Build Start`

I would suggest something like:
`Custom Message - All Statuses` - Matches the existing field

`Custom Message - Build Start` - new field if someone only wants to override the build start message

**Comment by timja** [comment]
@KreAch3R I've linked all the related issues in your description so that when this gets merged they will all get closed

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java [hunk 3]
```diff
@@ -458,6 +458,9 @@ public MessageBuilder appendCustomMessage(Result buildResult) {
                     customMessage = notifier.getCustomMessageFailure();
                 }
             }
+            if (customMessage.isEmpty()) {
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: can you also look at removing the check box from custom message so it is consistent with the other new fields?
thanks

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java (hunk 2 above)
Comment (KreAch3R) [reply src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: Thank you @timja 
You are right, it was a breaking change.

Check the latest commit I pushed. I think this restores the default behavior. The custom message will be applied to every build result, except if there is a specific custom message for that result.

File: pom.xml (hunk 1 above)
Comment (KreAch3R) [reply pom.xml]: I needed this for the build (using `mvn install`) to finish. I was getting the build error below:
https://del.dog/asinewohif.sql

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.20:test (default-test) on project slack: There are test failures.
...
[ERROR] Error occurred in starting fork, check output in log
[ERROR] Process Exit Code: 1
```

Build system: Windows 10 WSL, debian subsystem, openjdk-8-jdk java SDK

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java (hunk 2 above)
Comment (timja) [reply src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: What if a user wants a start message but none of the others?

File: pom.xml (hunk 1 above)
Comment (KreAch3R) [reply pom.xml]: I just confirmed this isn't related to this commit, I also need it to build the current master. Maybe I can push it as a different PR @timja, sorry about that

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java (hunk 2 above)
Comment (KreAch3R) [reply src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: I think I'm gonna leave it like that for this commit (so the default behavior remains the same, the custom message is applied everywhere). Your idea about the custom 'build start' and 'all statuses' is good though, I'll see if i can work on it in the future

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java (hunk 3 above)
Comment (KreAch3R) [reply src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: Which checkbox do you mean? the `includeCustomMessage` checkbox? This is applied to all custom messages, it works as expected afaik. Do you mean something else?

File: pom.xml (hunk 1 above)
Comment (timja) [reply pom.xml]: please, not keen to disable it globally, should be able to pass it on the command line though, I had a quick google and couldn't see anything

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java (hunk 3 above)
Comment (timja) [reply src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: you're right never mind

File: pom.xml (hunk 1 above)
Comment (KreAch3R) [reply pom.xml]: Ok, I will remove that part and also rebase to current master before force pushing into this PR to make the commit history cleaner, is that ok with you?

Comment (timja) [reply pom.xml]: sure sounds good, other wise good to merge after you get a green build on the rebase

Comment (KreAch3R) [reply pom.xml]: Ok, I found out that it's a bug in `openjdk-8-jdk`: https://stackoverflow.com/a/53016532/4008886
Debian bug: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=911925
[Workaround without involving maven pom.xmls](https://issues.apache.org/jira/browse/SUREFIRE-1588?focusedCommentId=16671777&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16671777)

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 4]
```diff
@@ -594,7 +656,13 @@ public SlackJobProperty(String teamDomain,
                                 boolean includeTestSummary,
                                 boolean showCommitList,
                                 boolean includeCustomMessage,
-                                String customMessage) {
+                                String customMessage,
```
Comment (casz) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]: Oh wow, I can't believe this made it through. adding stuff to a deprecated class feels wrong 😅 @timja @KreAch3R

Comment (timja) [reply src/main/java/jenkins/plugins/slack/SlackNotifier.java]: SlackNotifier isn't deprecated, an old constructor is deprecated that's being maintained for compatibility

**Comment by Wei-Lin-D** [comment]
I'm sorry but has this feature ever been released?
We are in version 2.3 but seems there still no ways to toggle on/off custom messages for each result.
How can we use this feature?
スクリーンショット 2019-10-03 19 07 12

**Comment by timja** [comment]
2.3 is about 2 years old I think... you can have a look at releases to see details

**Comment by Wei-Lin-D** [comment]
> 2.3 is about 2 years old I think... you can have a look at releases to see details

Thank you for reply!
While actually v2.30 was released on 16 Aug from the releases page...
And the latest version is v2.34 which was released 14 days ago.
This feature should be included in v2.30?

**Comment by timja** [comment]
You said you were using 2.3...
This feature was added ages ago, just use the latest version

**Comment by Wei-Lin-D** [comment]
> You said you were using 2.3...
> This feature was added ages ago, just use the latest version

Oh, my bad, I just got the version wrong (2.3 and 2.30) 😆 
Updated the plugin and now we can set custom messages for each build result.
Nice feature, thank you!

**Pull request merged**